sgkit/distance/api.py

```diff
@@ -77,27 +82,61 @@ def pairwise_distance(
 
     >>> x = np.array([[6, 4, 1,], [4, 5, 2], [9, 7, 3]])
     >>> pairwise_distance(x, metric='correlation')
-    array([[1.11022302e-16, 2.62956526e-01, 2.82353505e-03],
-           [2.62956526e-01, 0.00000000e+00, 2.14285714e-01],
-           [2.82353505e-03, 2.14285714e-01, 0.00000000e+00]])
+    array([[-4.44089210e-16,  2.62956526e-01,  2.82353505e-03],
+           [ 2.62956526e-01,  0.00000000e+00,  2.14285714e-01],
+           [ 2.82353505e-03,  2.14285714e-01,  0.00000000e+00]])
     """
-
     try:
-        metric_ufunc = getattr(metrics, metric)
+        metric_map_func = getattr(metrics, f"{metric}_map")
+        metric_reduce_func = getattr(metrics, f"{metric}_reduce")
+        n_map_param = metrics.N_MAP_PARAM[metric]
     except AttributeError:
         raise NotImplementedError(f"Given metric: {metric} is not implemented.")
 
     x = da.asarray(x)
-    x_distance = da.blockwise(
-        # Lambda wraps reshape for broadcast
-        lambda _x, _y: metric_ufunc(_x[:, None, :], _y),
-        "jk",
+    if x.ndim != 2:
+        raise ValueError(f"2-dimensional array expected, got '{x.ndim}'")
+
+    metric_param = np.empty(n_map_param, dtype=x.dtype)
+
+    def _pairwise(f: ArrayLike, g: ArrayLike) -> ArrayLike:
+        result: ArrayLike = metric_map_func(f[:, None, :], g, metric_param)
```

Comment:
Would it make sense to comment here why you need to add extra dimension?

Reply:
Now that I think about it, I think we don't really need this extra dimension. We can really simplify this by removing the extra dimension and making `combine` just return the chunk as the summation already happens in the metric's reduce function, this will also simplify the implementation and make is easy to grasp, let me change that and share in a bit. Let me know if you think otherwise.

Reply:
Ah, I see, combine makes sense for recursive aggregation, ignore the above comment in that case.